### PR TITLE
Time domain simulation

### PR DESCRIPTION
**Pull request by Abujazar**

### CODE REVIEW

File: femmt/component.py [hunk 1]
```diff
@@ -428,7 +429,7 @@ def set_winding_windows(self, winding_windows: list[WindingWindow]):
                         vww.turns.append(0)
 
         # Default values for global_accuracy and padding
-        self.mesh_data = MeshData(0.5, 1.5, mu_0, self.core.core_inner_diameter, self.core.window_w, self.windings)
+        self.mesh_data = MeshData(0.9, 1.5, mu_0, self.core.core_inner_diameter, self.core.window_w, self.windings)
```
Comment (gituser789) [inline femmt/component.py]: What is the reason for this mesh change?

File: femmt/component.py [hunk 2]
```diff
@@ -1823,6 +1856,148 @@ def visualize(self):
         gmsh.option.setNumber(f"View[{view}].NbIso", 40)
         view += 1
 
+        # if self.simulation_type == SimulationType.TimeDomain:
```
Comment (gituser789) [inline femmt/component.py]: is it possible to remove this big comment?

File: femmt/examples/get_avg_multiple_files.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import pandas as pd
```
Comment (gituser789) [inline femmt/examples/get_avg_multiple_files.py]: Is this file an example file?
The header and also the file name does not make clear that this is for time-domain simulation

is it possible to include this functionality into the toolbox? does it make sense?

File: femmt/component.py [hunk 4]
```diff
@@ -102,6 +106,14 @@ def __init__(self, simulation_type: SimulationType = SimulationType.FreqDomain,
         self.current = []                       # Defined for every conductor
         self.current_density = []               # Defined for every conductor
         self.voltage = []                       # Defined for every conductor
+        self.time = []                          # Defined for time domain simulation
+        self.average_currents = []              #Definded for average currents for every winding
+        self.rms_currents = []                  #definded for rms currents for every winding
+        self.delta_time = None
```
Comment (gituser789) [inline femmt/component.py]: comments for all new variables that they are for time-domain simulation

File: femmt/component.py [hunk 5]
```diff
@@ -732,7 +744,161 @@ def excitation(self, frequency: float, amplitude_list: List, phase_deg_list: Lis
...
+                    self.red_freq[num] = self.windings[num].strand_radius / self.delta
+                elif self.windings[num].conductor_type == ConductorType.RoundSolid:
+                    self.red_freq[num] = self.windings[num].conductor_radius / self.delta
+                else:
+                    ff.femmt_print("Reduced Frequency does not have a physical value here")
+                    ff.femmt_print(self.windings[num].conductor_type)
+                    self.red_freq[num] = 1  # TODO: doesn't make sense like this -> rewrite fore conductor windings shape
+        else:
+            # DC case
+            self.delta = 1e20  # random huge value
```
Comment (gituser789) [inline femmt/component.py:1062]: what is the variable 'delta' for? may rename the variable

File: femmt/component.py [hunk 6]
```diff
@@ -732,7 +744,161 @@ def excitation(self, frequency: float, amplitude_list: List, phase_deg_list: Lis
...
+            self.red_freq[num] = 0
+
+    def excitation_time_domain(self, frequency: float, current_list: List[List[float]], time_list: List[float], time_period: float, initial_time: float, timemax : float, NbSteps: int, delta_time: float, phase_deg_list: List = None, ex_type: str = 'current',
+                   plot_interpolation: bool = False, imposed_red_f=0):
+        """
+         Excites the electromagnetic problem in the time domain with specified current and time settings.
+
+        Parameters:
+        - frequency (float): The frequency of the excitation in Hz.
```
Comment (gituser789) [inline femmt/component.py:1072]: this text needs to go into the descriptions :param and :type
->
:param frequency: The frequency of the excitation in Hz
:type frequency: float

File: femmt/component.py [hunk 7]
```diff
@@ -864,15 +1033,59 @@ def single_simulation(self, freq: float, current: List[float], phi_deg: List[flo
...
+        # TODO: reuse center tapped
         if show_fem_simulation_results:
             self.visualize()
 
+    def time_domain_simulation(self, freq: float, current: List[List[float]], time: List[float], time_period: float, initial_time: float, timemax : float, NbSteps: int, delta_time: float,  phi_deg: List[float] = None,
+                          plot_interpolation: bool = False, show_fem_simulation_results: bool = True, show_rolling_average: bool = True, rolling_avg_window_size: int = 5 ):
+        """
+        Start a time_domain  electromagnetic ONELAB simulation.
+
+        :param plot_interpolation:
```
Comment (gituser789) [inline femmt/component.py:1363]: docstring not fully filled

File: femmt/component.py [hunk 8]
```diff
@@ -1772,6 +2083,411 @@ def calculate_and_write_log(self, sweep_number: int = 1, currents: List = None,
         with open(self.file_data.e_m_results_log_path, "w+", encoding='utf-8') as outfile:
             json.dump(log_dict, outfile, indent=2, ensure_ascii=False)
 
+    def calculate_and_write_log_try(self):
```
Comment (gituser789) [inline femmt/component.py]: what is the _try for?

File: femmt/component.py [hunk 9]
```diff
@@ -1772,6 +2083,411 @@ def calculate_and_write_log(self, sweep_number: int = 1, currents: List = None,
...
+                                                               log_dict["total_losses"]["all_windings_losses"]
+
+
+
+
+            #return log_dict
+
+
+
+        def common_log():
```
Comment (gituser789) [inline femmt/component.py]: missing docstirng

File: femmt/component.py [hunk 10]
```diff
@@ -1772,6 +2083,411 @@ def calculate_and_write_log(self, sweep_number: int = 1, currents: List = None,
...
+            log_dict["total_losses"]["eddy_core"] = sum(log_dict["single_sweeps"][d]["core_eddy_losses"] for d in range(len(log_dict["single_sweeps"])))
+            # For core losses just use hyst_losses of the fundamental frequency. When using single_simulation, the fundamental frequency is at [0]
+            # => just an approximation for excitation sweeps!
+            log_dict["total_losses"]["hyst_core_fundamental_freq"] = log_dict["single_sweeps"][fundamental_index]["core_hyst_losses"]
+
+            # Total losses of inductive component according to single or sweep simulation
+            log_dict["total_losses"]["core"] = log_dict["total_losses"]["hyst_core_fundamental_freq"] + log_dict["total_losses"]["eddy_core"]
+            log_dict["total_losses"]["total_losses"] = log_dict["total_losses"]["hyst_core_fundamental_freq"] + log_dict["total_losses"]["eddy_core"] + log_dict["total_losses"]["all_windings"]
+
+        def time_domain_log():
```
Comment (gituser789) [inline femmt/component.py]: missing docstring

File: femmt/component.py (hunk 8 above)
Comment (gituser789) [reply femmt/component.py]: it is not clear what is the difference between the sub-functions what can be found into this function. A small summary would be nice into the first doc-string

File: femmt/component.py [hunk 11]
```diff
@@ -2035,7 +2678,203 @@ def get_loss_data(self, last_n_values: int, loss_type: str = 'litz_loss'):
             data = list(reader)
         return data[-last_n_values:-1] + [data[-1]]
 
-    def load_result(self, res_name: str, res_type: str = "value", last_n: int = 1, part:str = "real", position: int = 0):
+    def get_rolling_average(self, window_size):
```
Comment (gituser789) [inline femmt/component.py:3159]: :param and :type are missing in the docstring

File: femmt/component.py [hunk 12]
```diff
@@ -2035,7 +2678,203 @@ def get_loss_data(self, last_n_values: int, loss_type: str = 'litz_loss'):
...
+        """
+        Computes the rolling average for each result file and plots the original data
+        alongside the rolling average if show_rolling_average is True.
+        """
+
+        def rolling_calculation(res_name: str, res_path: str):
```
Comment (gituser789) [inline femmt/component.py:3165]: docstring missing

File: femmt/examples/basic_transformer.py [hunk 13]
```diff
@@ -81,17 +81,27 @@ def example_thermal_simulation(show_visual_outputs: bool = True, flag_insulation
         os.mkdir(working_directory)
 
     # 1. chose simulation type
-    geo = fmt.MagneticComponent(component_type=fmt.ComponentType.Transformer, working_directory=working_directory,
+    geo = fmt.MagneticComponent(simulation_type=fmt.SimulationType.FreqDomain, component_type=fmt.ComponentType.Transformer, working_directory=working_directory,
```
Comment (gituser789) [inline femmt/examples/basic_transformer.py]: i do not agree with changing this file in general.
There should be a new file called basic_transformer_time_domain.py to show the time domain simulation for the transformer

Comment (gituser789) [reply femmt/examples/basic_transformer.py]: ok, found the file some lines down.
Nevertheless, this file should not be changed.

**Review by gituser789 — CHANGES_REQUESTED** [review]

File: femmt/component.py (hunk 1 above)
Comment (Abujazar) [reply femmt/component.py]: to reduce the simulation's duration. It takes long time with more mesh accuracy

**Review by gituser789 — APPROVED** [review]

**Pull request merged**